src/data-access/services/basic-ftp-ftp-client.ts

```diff
@@ -0,0 +1,89 @@
+import { FtpClient } from '../interfaces/ftp-client'
+import { FtpConnectionConfiguration } from '../value-objects/ftp-connection-configuration'
+import * as basicFtp from 'basic-ftp'
+import { PassThrough } from 'node:stream'
+import { Logger } from '../../logger/logger'
+import { FileMetadata } from '../value-objects/file-metadata'
+import { ConnectionStatus } from '../enums/connection-status'
+
+const CONNECTION_TIMEOUT_DURATION_IN_MS: number = 4000
+
+export class BasicFtpFtpClient implements FtpClient {
+  private onConnectionStatusChangedCallback?: (connectionStatus: ConnectionStatus) => void
+  private readonly ftpClient: basicFtp.Client = new basicFtp.Client(CONNECTION_TIMEOUT_DURATION_IN_MS)
+  private readonly logger: Logger
+
+  public constructor(
+    private readonly configuration: FtpConnectionConfiguration,
+    logger: Logger,
+  ) {
+    this.logger = logger.tag(this.constructor.name)
+  }
+
+  public async connect(): Promise<void> {
+    try {
+      this.onConnectionStatusChangedCallback?.(ConnectionStatus.CONNECTING)
+      await this.ftpClient.access({
+        host: this.configuration.host,
+        port: this.configuration.port,
+        user: this.configuration.user,
+        password: this.configuration.password,
+        secure: false,
+      })
+      this.onConnectionStatusChangedCallback?.(ConnectionStatus.CONNECTED)
+      this.logger.info(`Connected to FTP server ${this.configuration.host}:${this.configuration.port}.`)
+      this.ftpClient.ftp.socket.on('timeout', () => {
+        this.logger.info(`The connection to the FTP server ${this.configuration.host}:${this.configuration.port} timed out.`)
+        this.onConnectionStatusChangedCallback?.(ConnectionStatus.DISCONNECTED)
+      })
+      this.ftpClient.ftp.socket.on('end', () => {
+        this.logger.info(`FTP server ${this.configuration.host}:${this.configuration.port} terminated the connection.`)
+        this.onConnectionStatusChangedCallback?.(ConnectionStatus.DISCONNECTED)
+      })
```

Comment:
I tried my hand at some code i find more fun to look at 😄  What do you think?

```suggestion
 this.onConnectionStatusChangedCallback?.(ConnectionStatus.CONNECTING)
      const { host, port, user, password } = this.configuration

      await this.ftpClient.access({ host, port, user, password, secure: false }).then(() => {
        this.onConnectionStatusChangedCallback?.(ConnectionStatus.CONNECTED)
        this.logger.info(`Connected to FTP server ${host}:${port}.`)
      })

      this.ftpClient.ftp.socket.on('timeout', () => {
        this.logger.info(`The connection to the FTP server ${host}:${port} timed out.`)
        this.onConnectionStatusChangedCallback?.(ConnectionStatus.DISCONNECTED)
      })

      this.ftpClient.ftp.socket.on('end', () => {
        this.logger.info(`FTP server ${host}:${port} terminated the connection.`)
        this.onConnectionStatusChangedCallback?.(ConnectionStatus.DISCONNECTED)
      })
```

Reply:
I have added some vertical spacing and extracted the bulk into separate methods to (hopefully) make it more readable :)